File: solved/2294.cpp
#include <bits/stdc++.h>
using namespace std;

int arr[101];
int arr2[101];
int dp[10001][101] = {};
int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  int n, k;
  cin >> n >> k;
  for (int i = 0; i < n; i++) cin >> arr[i];
  sort(arr, arr + n);
  arr2[0] = arr[0];
  int count = 1;
  for (int i = 1; i < n; i++) {
    if (arr[i - 1] == arr[i]) continue;
    arr2[count++] = arr[i];
  }
  for (int i = 0; i <= k; i++) {
    for (int j = 0; j < n; j++) {
      dp[i][j] = 123456789;
    }
  }
  for (int i = 0; i < n; i++) {
    if (arr[i] <= k) dp[arr[i]][i] = 1;
  }

  for (int i = 0; i <= k; i++) {
    for (int j = 0; j < n; j++) {
      if (arr[j] > i) break;
      for (int l = 0; l <= j; l++) {
        dp[i][j] = min(dp[i][j], dp[i - arr[j]][l] + 1);
      }
    }
  }

  int ans = dp[k][0];
  for (int i = 0; i < n; i++) {
    ans = min(ans, dp[k][i]);
  }
  if (ans >= 123456789) ans = -1;
  cout << ans << endl;

  return 0;
}

File: solved/2294.py
import sys


n, k = map(int, sys.stdin.readline().strip().split())
arr = []
for i in range(n):
    arr.append(int(sys.stdin.readline().strip()))
arr = list(set(arr))
arr.sort()
n = len(arr)
# print(arr)

# dp[i][j] = 현재 채운 돈 i, 마지막에 arr[j] 동전 사용
dp = []
for i in range(k+1):
    dp.append([123456789]*n)

for i in range(n):
    if arr[i]<=k:
        dp[arr[i]][i] = 1

for i in range(k+1):
    for j in range(n):
        if arr[j] > i:
            break
        for l in range(j+1):
            dp[i][j] = min(dp[i][j], dp[i-arr[j]][l]+1)

# print(dp)
ans = min(dp[k])
if ans >= 123456789:
    ans = -1
print(ans)
